src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,271 @@
+
+  visibleBiasNegativeGrad = arma::mat(
+      negativeGradient.memptr() + weightNegativeGrad.n_elem +
+      hiddenBiasNegativeGrad.n_elem,
+      numVisible, 1, false, false);
+
+  weightPositiveGrad = arma::mat(positiveGradient.memptr(),
+    visible.Weight().n_rows,
+    visible.Weight().n_cols, false, false);
+
+  hiddenBiasPositiveGrad = arma::mat(
+    positiveGradient.memptr() + weightPositiveGrad.n_elem,
+    numHidden, 1, false, false);
+  
+  visibleBiasPositiveGrad = arma::mat(
+      positiveGradient.memptr() + weightPositiveGrad.n_elem +
+      hiddenBiasPositiveGrad.n_elem,
+      numVisible, 1, false, false);
+
+  gibbsTemporary.set_size(numHidden, 1);
+  negativeSamples.set_size(numVisible, 1);
+  preActivation.set_size(numHidden, 1);
+  corruptInput.set_size(numVisible, 1);
+
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+template<template<typename> class Optimizer>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Train(const arma::mat& predictors, Optimizer<NetworkType>& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+  {
+    Reset();
+  }
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+ /* 
+  * This function calculates
+  * the free energy of the model
+  */
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    FreeEnergy(arma::mat&& input)
+{
+  visible.ForwardPreActivation(std::move(input), std::move(preActivation));
+  SoftplusFunction::Fn(preActivation, activation);
+  return  -(arma::accu(activation) + arma::dot(input, visible.Bias()));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Evaluate(const arma::mat& /* parameters*/, const size_t i)
+{
+  corruptInput = arma::round(predictors.col(i));
+  if (!useMonitoringCost)
+  {
+    // Do not use unsafe col predictor as we change the input
+    Gibbs(std::move(predictors.col(i)), std::move(negativeSamples));
+    return (FreeEnergy(std::move(predictors.col(i))) - 
+        FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      size_t idx = RandInt(0, predictors.n_rows);
+      corruptInput.row(idx) = 1 - corruptInput.row(idx);
+      return std::log(LogisticFunction::Fn(FreeEnergy(std::move(corruptInput)) -
+          FreeEnergy(std::move(arma::round(predictors.col(i)))))) *
+          predictors.n_rows;
+    }
+    else
+    {
+      // Cross entropy loss
+      visible.Forward(std::move(predictors.col(i)),
+        std::move(activation));
+      return arma::accu( predictors.col(i) * arma::log(activation).t() +
```

Comment:
Well the activation can't be greater than 1 as it is output of the sigmoid unit whose range is 0 to 1. I will dig deeper to find the problem.